### ISSUE TITLE
Implement drop_region() for StorageEngine

### ISSUE THREAD

**Issue by evenyag** [GreptimeTeam/greptimedb#1556]
### What problem does the new feature solve?

This is a necessary feature to solve #497.

The StorageEngine should support dropping a region, so we could clean files and other resources occupied by that region. This can also affect the WAL purge task since the WAL of a dropped region will never be compacted.

### What does the feature do?

Dropping a region needs to remove all the WAL and SSTs related to that region and finally remove the region from the StorageEngine. This also releases all the memtables.

Since we might have queries running while dropping a region, we couldn't remove the files immediately. So we rely on the FilePurger to remove those files.

Errors can occur while dropping a region so we should persist the message that the region is dropped in the manifest.

### Implementation challenges

drop_region might perform the following steps:
- acquires the write lock
- persists a DropAction to the manifest
- deletes WAL
- marks all SSTs as deleted
- tells the background gc task to remove all the manifest when all SSTs are deleted

When the storage engine reopens the region and finds the region is deleted (via the DropAction in the manifest), it should continue the dropping process

**Comment by DevilExileSu** [comment]
Hi @evenyag , If possible, I would like to try this.

**Comment by evenyag** [comment]
We can break this into several small tasks. I plan to provide the simplest `drop_region()` implementation that only deletes WAL as the first step. But I haven't put it into my schedule yet so I think I won't work on this in the last two weeks.

If you don't mind, you could try the first step. @DevilExileSu

**Comment by DevilExileSu** [comment]
> We can break this into several small tasks. I plan to provide the simplest `drop_region()` implementation that only deletes WAL as the first step. But I haven't put it into my schedule yet so I think I won't work on this in the last two weeks.
> 
> If you don't mind, you could try the first step. @DevilExileSu

Thank you for your reply! Is the deletion of WAL to be logically deleted through [LogStore::delete_namespace](https://github.com/GreptimeTeam/greptimedb/blob/7a9dd5f0c8a89443d261c20787a3cc3be73e2c44/src/log-store/src/raft_engine/log_store.rs#L240)? It seems that the WAL file cannot be deleted directly.

**Comment by evenyag** [comment]
> Is the deletion of WAL to be logically deleted through [LogStore::delete_namespace](https://github.com/GreptimeTeam/greptimedb/blob/7a9dd5f0c8a89443d261c20787a3cc3be73e2c44/src/log-store/src/raft_engine/log_store.rs#L240)?

Yes. You should also call `LogStore::obsolete()` to mark all log entries as obsolete first. We should do it inside the `RegionWriter`. One approach is adding an `on_drop` method to the writer, acquiring the write lock, and removing data of the WAL. Its implementation is similar to close() but does more work.
https://github.com/GreptimeTeam/greptimedb/blob/7a9dd5f0c8a89443d261c20787a3cc3be73e2c44/src/storage/src/region/writer.rs#L251-L272

**Comment by DevilExileSu** [comment]
I tried to implement `on_drop` in `RegionWriter` by following steps
1. Acquires the write lock.
2. Close writer reject any potential writing.
3. Waits or cancels the flush job.
4. TODO: Waits or cancels the Compaction Task
5. Mark all data obsolete in the WAL.
6. Add `RegionMetaAction::Remove` to recover from manifest in case of failure. The main task is to restore the cleaning of sst files. If there is a failure in the previous steps, it can be restored through the `Procedure` framework.
7. Delete the namespace of the region from the WAL.
8. Get all sst files for current version.
9. Trigger sst files purger task through `LevelMetas::merge()`

In the current implementation, if the sst files are deleted by using the background GC task (`FilePurger`) method, it seems that it will not open the Manifest of the dropped region to process the unfinished GC task when recovering from a failure.

**Comment by evenyag** [comment]
Oops, sorry for the late reply.

Some suggestions:

> Trigger sst files purger task through LevelMetas::merge()

You could a new method that just marks all the files are deleted. This is different from the merge operation.


> * Mark all data in WAL as obsolete.
> * Add `RegionMetaAction::Remove` to recover from manifest in case of failure.

Better to store the `Remove` meta action before deleting data in WAL

> if the sst files are deleted by using the background GC task (`FilePurger`) method, it seems that it will not open the Manifest of the dropped region to process the unfinished GC task when recovering from a failure.

We could remove all the SST files (delete them explicitly or mark them as deleted again) when we open a region and find out the region is dropped via the remove meta action.

Currently, files are deleted in an async manner (via FilePurger). They usually don't block the drop operation so we could remove the manifest before all files are deleted. However, if the process restarts before the file purger removes all the files. Those files become dangling files. We need a GC task to find out dangling files and clean them.

The GC task could be tracked by another issue and doesn't block the implementation of `drop_region()`.

For simplicity, you could only perform the following steps in the first implementation
1. Acquires the write lock.
2. Close writer reject any potential writing.
3. Waits or cancels the flush job.
4. Mark all data in WAL as obsolete.
5. Delete the namespace of the region from the WAL.
6. Mark SSTs as deleted.
7. Remove the region from StorageEngine's region map.

In the next PR, you could add the remove meta action before step 3 and remove all the manifests after step 6

**Comment by DevilExileSu** [comment]
Next plan:
1. Improve [drop procedure](https://github.com/GreptimeTeam/greptimedb/blob/develop/src/mito/src/engine/procedure/drop.rs) to update table manifest.
2. Implement a `StorageManager`(or another suitable name) in `Datanode`, like [LocalManager](https://github.com/GreptimeTeam/greptimedb/blob/783a794060178df4897016970df3040e17e0714d/src/common/procedure/src/local.rs#L366-L372), to create regular scan tasks.
3. Regularly scan the storage. Utilize `CatalogManagerRef` to scan the directory where the table is located and read the table manifest. This will help determine if the table has been removed or identify any orphan files. If the table is removed, use bg task to purge the associated files.
4. Refactored [FilePurgeHandler](https://github.com/GreptimeTeam/greptimedb/blob/783a794060178df4897016970df3040e17e0714d/src/storage/src/file_purger.rs#L42) and [FilePureRequest](https://github.com/GreptimeTeam/greptimedb/blob/783a794060178df4897016970df3040e17e0714d/src/storage/src/file_purger.rs#L26-L30) to be more generic, not just for purging parquet files.

I would greatly appreciate any suggestions, feedback, or ideas you may have regarding these planned improvements.
